### ISSUE TITLE
amp-inline-gallery: Status messages are not explicitly announced by assistive technologies

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#31176]
## WCAG Level
Level AA

## Priority
Medium

## Type
Semantics

## Pages/screens/components affected
* [amp-inline-gallery](https://amp.dev/documentation/components/amp-inline-gallery/)
* [amp-lightbox-gallery](https://amp.dev/documentation/components/amp-lightbox-gallery/)

## Description
When status messages/notifications appear on a page dynamically, and without automatically receiving keyboard focus, it is important that these messages are also communicated correctly to assistive technologies.

The `<amp-inline-gallery>` (which is also used for [one of the `amp-lightbox-gallery` examples](https://nutritious-tide-hornet.glitch.me/amp-inline-gallery/basic.html)) includes the ability to have a visible `<amp-inline-gallery-pagination>` to indicate the current slide and total slides in the gallery. This is dynamically updated as the slides are changes, but such changes are not announced by screen readers. In addition, the entire pagination element is hidden from assistive technologies using `aria-hidden="true"`.

![The pagination indicator on the amp-lightbox-gallery example, showing a '2/21' indicator](https://user-images.githubusercontent.com/54933393/99181998-fc999c80-2729-11eb-8093-d6086c5d83ff.png)

Note that when there are eight or less total slides, the pagination element only displays a series of dots underneath the carousel.

![The pagination indicator on the amp-inline-gallery example, displayed as dots](https://user-images.githubusercontent.com/54933393/99182001-03281400-272a-11eb-8e44-64d0c6281957.png)

## User impact
People using assistive technologies such as screen readers and screen magnification tools may not be aware when status messages are injected into the page outside of their current area of focus. For example, if a confirmation message only appears visually (without being explicitly announced) as a "toast" style notification, or even as a permanently visible new notification block in a specific part of the screen that is not currently in focus, screen reader users may not even be aware that a confirmation message appeared at all.

In the case of the inline galleries, users will not be made aware of how many slides in total are present in the gallery, and which slide they are currently viewing.

## Required solution
Whenever status messages or notifications are injected into the page, and these status messages do not receive focus (which is usually not appropriate, since this moves the user's context and interrupts their current activity), make sure that the messages are coded in such a way that they are exposed to, and consequently conveyed by, assistive technologies. For web-based content, this typically means identifying the status message as an [ARIA live region](https://www.w3.org/TR/wai-aria-1.1/#attrs_liveregions) - by either using a [live region `role`](https://www.w3.org/TR/wai-aria-1.1/#live_region_roles), or adding an [`aria-live` attribute](https://www.w3.org/TR/wai-aria-1.1/#aria-live).

This solution must be applied to all instances of the issue identified within the test sample, then applied to all other instances of the same issue identified throughout the rest of the website.

## Implementation guidance

Currently, the pagination element results in the following markup structure.

```
<amp-inline-gallery-pagination layout="nodisplay" inset="" class="i-amphtml-layout-nodisplay i-amphtml-element i-amphtml-built i-amphtml-layout" hidden="" i-amphtml-layout="nodisplay">
  <div class="i-amphtml-inline-gallery-pagination-container" aria-hidden="true">
    <div class="i-amphtml-inline-gallery-pagination-dots">
      ...
    </div>
    <div class="i-amphtml-inline-gallery-pagination-numbers" hidden="">
      <div class="i-amphtml-inline-gallery-pagination-frosting"></div>
      <div class="i-amphtml-inline-gallery-pagination-backdrop"></div>
      <div class="i-amphtml-inline-gallery-pagination-background"></div>
      <div class="i-amphtml-inline-gallery-pagination-count">
        <span class="i-amphtml-inline-gallery-pagination-index">1</span> <span>/ </span>
        <span class="i-amphtml-inline-gallery-pagination-total">3</span>
      </div>
    </div>
  </div>
</amp-inline-gallery-pagination>
```

Instead of using `aria-hidden="true"` to hide entirety of the pagination element, we recommend moving this attribute just to the "dots" container.

For situations where there are eight or less slides, and the pagination counter is not displayed, remove the `hidden` attribute on the container for the pagination numbers (which results in the element being fully hidden, not only visually, but also from assistive technologies), and instead use CSS to visually hide the control while keeping its content accessible (for instance, using a `.visually-hidden` class).

```
.visually-hidden {
  clip: rect(0 0 0 0);
  clip-path: inset(50%);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}
```

Turn the container into an ARIA live region by adding `aria-live="polite"` (and, to treat it as a single entity, `aria-atomic="true"`). Any dynamic changes inside the container will then be announced by assistive technologies (and due to the `aria-atomic` attribute, the entire component's content will be announced, rather than just the one element that has been updated).

```
<amp-inline-gallery-pagination layout="nodisplay" inset="" class="i-amphtml-layout-nodisplay i-amphtml-element i-amphtml-built i-amphtml-layout" hidden="" i-amphtml-layout="nodisplay">
  <div class="i-amphtml-inline-gallery-pagination-container">
    <div class="i-amphtml-inline-gallery-pagination-dots" aria-hidden="true">
      ...
    </div>
    <div class="i-amphtml-inline-gallery-pagination-numbers visually-hidden">
      <div class="i-amphtml-inline-gallery-pagination-frosting"></div>
      <div class="i-amphtml-inline-gallery-pagination-backdrop"></div>
      <div class="i-amphtml-inline-gallery-pagination-background"></div>
      <div class="i-amphtml-inline-gallery-pagination-count" aria-live="polite" aria-atomic="true">
        <span class="i-amphtml-inline-gallery-pagination-index">1</span> <span>/ </span>
        <span class="i-amphtml-inline-gallery-pagination-total">3</span>
      </div>
    </div>
  </div>
</amp-inline-gallery-pagination>
```

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied to issues identified within the test sample, and to test the rest of the website for instances of the same issue:
1. Turn on your screen reader, and open a page that includes a status message.
2. Activate the component that triggers the status message.
3. Make sure that screen readers announce appropriate feedback when the status message visibly appears on the page.

### Definition of done
Complete all of these tasks before closing this issue or indicating it is ready for retest:
* All issues identified within the test sample have been resolved.
* The rest of the website has been tested for the same issue.
* All issues identified throughout the website have been resolved or filed as new issues.

## Related standards
[WCAG 2.1 Success Criterion 4.1.3 Status Messages (Level A)](https://www.w3.org/TR/WCAG21/#status-messages)

## More information
* [W3C - Understanding Success Criterion 4.1.3: Status Messages](https://www.w3.org/WAI/WCAG21/Understanding/status-messages.html)
* [Knowbility - Exploring WCAG 2.1: 4.1.3 Status Messages](https://knowbility.org/blog/2018/WCAG21-413StatusMessages/)
* [MDN web docs - ARIA live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions)
* [How-to: Hide content / The A11y Project](https://www.a11yproject.com/posts/2013-01-11-how-to-hide-content/)

## Test data
Test date: November 2020
